Add two LCAO base group GPU version compilation options in toolchain

tang070205 opened the pull request:
Firstly, it's cusolverMp merhod

Change the link libraries for the cal and cusolverMp sections in the CMakeLists.txt file to manually specified, and then manually pass the parameters in `cmake -D CAL_CUSOLVERMP_PATH=/path/to/lib`

Add the following options after build_abacus_gnu.sh
`-DUSE_CUDA=ON \`
`-DENABLE_CUSOLVERMP=ON \`
`-D CAL_CUSOLVERMP_PATH=/opt/nvidia/hpc_sdk/Linux_x86_64/24.11/math_libs/12.6/targets/x86_64-linux/lib`
CAL_CUSOLVERMP-PATH needs to be set according to different environments

next it is necessary to set up the environment for cal and hpcx, which can be added to~/. bashrc or manually set up an env.sh
`export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/nvidia/hpc_sdk/Linux_x86_64/24.11/comm_libs/12.6/hpcx/hpcx-2.20/ucc/lib`
`export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/nvidia/hpc_sdk/Linux_x86_64/24.11/comm_libs/12.6/hpcx/hpcx-2.20/ucx/lib`
`export CPATH=$CPATH:/opt/nvidia/hpc_sdk/Linux_x86_64/24.11/math_libs/12.6/targets/x86_64-linux/include`


The second is the ELPA method

Add the following options after toolchain_gnu.sh
`export CUDA_PATH=/usr/local/cuda`
`--enable-cuda \`
`--gpu-ver=89 \`
The 40 Series here is newly added in the install_abacus_toolchain.sh file, corresponding to sm_89


The above two methods can be compiled successfully by using the build_abacus_gnu.sh file

tang070205 committed:
Add optional LCAO base GPU versions supported by cusolvermp

tang070205 committed:
Add optional LCAO base GPU versions supported by elpa

tang070205 committed:
Add optional LCAO base GPU versions supported by elpa

tang070205 committed:
Add L40S as GPUVER value for sm_89 architecture

tang070205 committed:
Delete a few lines of content to enable Nvidia to compile

tang070205 committed:
Add a specified Fortran mpi compiler for elpa to use

tang070205 committed:
Add CUDA path for use by ELPA-GPU

tang070205 committed:
Add optional LCAO base GPU versions supported by elpa

tang070205 committed:
Modify a small issue

tang070205 committed:
Change to manually specifying the link libraries for CAL and cusolverMp

tang070205 committed:
Merge branch 'develop' into develop

tang070205 committed:
Add the use of 'cusolvermp' or 'elpa' methods to compile ABACUS GPU-LCAO

tang070205 committed:
Add the use of 'cusolvermp' or 'elpa' methods to compile ABACUS GPU-LCAO

tang070205 committed:
Add the use of 'cusolvermp' or 'elpa' methods to compile ABACUS GPU-LCAO

QuantumMisaka commented:
I'll submit a modification. After that, the ELPA-GPU compilation can be smoothly done in any cuda>11.6 environment. the installation and usage of cusolvermp needs more development

QuantumMisaka committed:
Add modification
- ELPA compiler flags modification
- GPU_VER setting modification: user should specify the GPU compability number, but not the GPU name
- Modify toolchain_[gnu,intel].sh and build_abacus_[gnu,intel].sh to use the above modification

QuantumMisaka committed:
minor adjustment

QuantumMisaka committed:
update README

QuantumMisaka committed:
give back cmake default option

tang070205 committed:
Merge pull request #1 from QuantumMisaka/lcao-gpu-modify

LCAO-GPU modification

QuantumMisaka commented:
We will try to add cusolvermp installation & compilation inside toolchain

QuantumMisaka commented:
> We will try to add cusolvermp installation & compilation inside toolchain

Update: The method in deploying cusovermp and the related dependencies （UCC/UCX/libcal）can be multiple (nvhpc-sdk/individual-package) , and different version of these package may have different path, making it very difficult to link them automatcally. Besides, if one wants to use the multiple-GPU calculation, simplily install HPC_SDK automatically is not enough, and the UCC/UCX/cusolvermp need to be complied/deployed according to the server setting. So, we will not add automatically download/link in toolchain now. However, we will add a simple deployment method via HPC-SDK in README for user as reference to install and use the cusolvermp themselves.

QuantumMisaka committed:
update README and cusolvermp

tang070205 committed:
Merge pull request #2 from QuantumMisaka/lcao-gpu-modify

update README and cusolvermp

QuantumMisaka reviewed (approved):
All tests passed. cusolvermp is too complicated to be incorporate

QuantumMisaka commented on toolchain/README.md:
````diff
@@ -187,6 +190,74 @@ or you can also do it in a more completely way:
+2. For the cusolvermp method, toolchain_*.sh does not need to be changed, just follow it directly install dependencies using `./toolchain_*.sh`, and then add
+```shell
+-DUSE_CUDA=ON \
+-DUSE_CUSOLVERMP=ON \
````
Should be -DENABLE_CUSOLVERMP=ON

tang070205 committed:
Update README.md

QuantumMisaka commented:
@mohanchen @dzzz2001 @goodchong We think this version of toolchain is enough now for a simple GPU-LCAO installation with CUDA and ELPA

mohanchen commented:
LGTM
